manager_42.py

```diff
@@ -32,55 +32,47 @@
 import sys
 import time
 import yaml
+from collections import namedtuple
 
 from osclib.memoize import memoize
 
 logger = logging.getLogger()
 
-OPENSUSE = 'openSUSE:Leap:42.3'
-
 makeurl = osc.core.makeurl
 http_GET = osc.core.http_GET
 http_DELETE = osc.core.http_DELETE
 http_PUT = osc.core.http_PUT
 http_POST = osc.core.http_POST
 
 class Manager42(object):
-    def __init__(self, from_prj, caching = True):
-        self.from_prj = from_prj
+    def __init__(self, caching = True, configfh = None):
         self.caching = caching
         self.apiurl = osc.conf.config['apiurl']
-        self.project_preference_order = [
-                'SUSE:SLE-12-SP3:GA',
-                'SUSE:SLE-12-SP2:Update',
-                'SUSE:SLE-12-SP2:GA',
-                'SUSE:SLE-12-SP1:Update',
-                'SUSE:SLE-12-SP1:GA',
-                'SUSE:SLE-12:Update',
-                'SUSE:SLE-12:GA',
-                'openSUSE:Leap:42.2:Update',
-                'openSUSE:Leap:42.2',
-                'openSUSE:Leap:42.2:NonFree:Update',
-                'openSUSE:Leap:42.2:NonFree',
-                'openSUSE:Leap:42.1:Update',
-                'openSUSE:Leap:42.1',
-                'openSUSE:Leap:42.1:NonFree:Update',
-                'openSUSE:Leap:42.1:NonFree',
-                'openSUSE:Factory',
-                'openSUSE:Factory:NonFree',
-                'openSUSE:Leap:42.3:SLE-workarounds',
-                'openSUSE:Leap:42.2:SLE-workarounds',
-                ]
-
-        self.parse_lookup(self.from_prj)
+
+        self.config_defaults = {
+            'project_preference_order' : [],
+            'from_prj' : 'openSUSE:Leap:42.3',
+            'factory' : 'openSUSE:Factory',
+            }
+
+        self.config = self._load_config(configfh)
```

Comment:
Presumably the config file changing while a process is running and instantiating multiple `Manager42`/`ReviewBot` instances is just a waste. As such would it make more sense to load the config once as a class variable instead of parsing for each new instance? I did this when I incorporated your config patch into #614 (see https://github.com/openSUSE/osc-plugin-factory/pull/614/commits/1fc8ce3905f20cbf3afe28159a6ae63285d2b1d2).

Given that you had used this same style for a potential `ReviewBot` config, does it make more sense to abstract as a separate config utility that both can use? In the same fashion as `osc.conf`. Although after writing that I remembered `osclib/conf.py` which populates `osc.conf`. Either way I imagine this code might be better to be shared?

And I see `# FIXME: add to ToolBase and rebase Manager42 on that`. Alright. I only saw `biarchtool.py` using `ToolBase` so I didn't use for #628 , should I rebase on that as well? If this is to be used by `ReviewBot` should it be rebased on `ToolBase` or perhaps just call a static method to retrieve config?